### PR TITLE
Add support for email address changes.

### PR DESCRIPTION
**Pull request by demiankatz**
Like #1367, this PR is designed to resolve a prerequisite for [VUFIND-1309](https://vufind.org/jira/browse/VUFIND-1309), though this is also a useful feature in its own right. See also #1354 for some earlier discussion about this feature.

TODO
- [x] Get initial implementation working
- [x] Add support for appropriate interaction with email verification functionality
- [x] Confirm CSRF protection is appropriately applied
- [x] Confirm CAPTCHA support is fully functional
- [x] Add flag to differentiate between user-specified change and upstream change; do not overwrite user-specified changes.
- [x] Review how to handle email propagation to auth handler (and whether to block email change when auth handler does not support updates).
- [x] Test notification to old email address of changes.
- [x] Test with PostgreSQL
- [x] Write Mink test to cover new form
- [x] Update changelog (including database changelog)
- [x] Run full test suite

### CODE REVIEW

**Comment by demiankatz** [comment]
I've requested a couple of reviews on this. @EreMaijala, I'm interested to hear how this compares to whatever functionality you have implemented in Finna (I did not attempt to investigate your code, since I figured this would be easy enough to build from the ground up, and the verification concerns were probably different anyway). @crhallberg, I'm interested in your opinion on the new form and button.

I'd also be interested to hear from @mtrojan-ub in case he has any different opinions on how the verification functionality should be handled.

Thanks, everyone!

**Comment by demiankatz** [comment]
I should also note that there are some test-related items still left on the to-do list; I am going to hold off on investing time in these until I am sure we are all on the same page about basic functionality. Once we're in agreement about how to proceed, I will go ahead with those finishing touches.

File: module/VuFind/src/VuFind/Controller/MyResearchController.php [hunk 1]
```diff
@@ -1794,6 +1807,76 @@ public function newPasswordAction()
...
+                    $this->sendVerificationEmail($user, true);
+                } else {
+                    $this->flashMessenger()
+                        ->addMessage('new_email_success', 'success');
+                }
+            } catch (AuthException $e) {
+                $this->flashMessenger()->addMessage($e->getMessage(), 'error');
+                return $view;
+            }
+            // Go to favorites
```
Comment (EreMaijala) [inline module/VuFind/src/VuFind/Controller/MyResearchController.php]: "Go to MyResearch default page"?

**Comment by EreMaijala** [comment]
For us it's a bit problematic that the email stored in VuFind is so tightly linked to the one in the authentication backend. We actually treat e.g. the email address coming from an ILS just a default and allow the user to change it separately in the ILS and in VuFind. This allows the user to have a different email address for whatever use they want in VuFind while keeping the "official" one in ILS. Also, with e.g. Shibboleth there'll never be a way for the authentication backend to allow email address change.

This proposal makes sense in the current context and how VuFind always overwrites the email address with anything coming from the auth method, but I believe it would be beneficial to not necessarily link the two emails so tightly. I also think it should be considered in a wider context for future functionality regarding updates to the ILS profile. We already support updating of user details in the ILS with Koha, and I suppose similar functionality would be possible for other backends too. If not directly via the ILS driver, possibly by other means such as storing the change request in VuFind's database and alerting administrators about it.

File: module/VuFind/src/VuFind/Controller/MyResearchController.php (hunk 1 above)
Comment (demiankatz) [reply module/VuFind/src/VuFind/Controller/MyResearchController.php]: Good catch -- that's what I get for copying and pasting. I've fixed not only this instance, but also the one I copied here. :-)

**Comment by demiankatz** [comment]
@EreMaijala, you make good points. I suppose the more "correct" way to do this would be to have a user_email table with user_id, email_address, verified_status, and then put a preferred_email_id as a foreign key in the user table. There could also potentially be a setting to control how many addresses a single user could have on file... and we'd have to make a decision about whether to allow non-unique addresses in the database (I think the answer would likely be yes -- we don't want to create a system where one user can lock another one out by abusing the email functionality... but this might have implications for password recovery, etc.).

Given that email is used for very little in the code right now, this refactor would not be terribly difficult to accomplish. As you say, the most complex work would probably be figuring out how to adapt the current logic that saves email addresses into the user table to do something a little more sophisticated.

Then it just becomes a question of strategy: do we want to do all of that work now, or do we want to stick with this simpler version as a "minimum viable product" so we can move forward for 6.1, and then do the bigger job as part of VuFind 7?

I'm a bit torn about this. I think redesigning the email system is too substantial of a change for a point release (you really can't argue that it's a new feature so much as a redesign, especially since it will require some significant data migration), so it would have to wait until version 7 to get merged. That seems to argue for going forward with this simpler approach for 6.1... but then the problem is that we'll end up introducing a new feature and almost immediately reversing course on its implementation, which could be confusing for everyone (both in terms of code and UI).

My ultimate goal is to take care of [VUFIND-1309](https://vufind.org/jira/browse/VUFIND-1309), which is only tangentially related to this... but finding some kind of solution to this problem is necessary to make that work viable.

I welcome further thoughts -- and we can also discuss this further on the dev call later this morning!

**Comment by demiankatz** [comment]
...mulling this over a little further, I suppose if we did away with the idea of preferred_email_id and just used the existing email field in user to set the preferred email address (sort of similar to the way library cards work) that would make the change significantly less disruptive. There would still need to be some changes on the verification side -- but maybe it could be simplified enough to make this workable without dramatically changing the code. I'll keep thinking about this and we can discuss further in a little while....

**Comment by EreMaijala** [comment]
I suppose in a perfect world the email table would be used, but I also think it doesn't need to be that complicated. What we do:

* Upon logging in, email address from the authentication system is only added to the user table if there's no existing email address. This could be controlled with the change_email setting.
* User can change the email address stored in VuFind's database. This is where the pending_email column would be useful since we're lacking verification and any typos will cause errors with SDI etc.
* If supported, user can change (or request change of) the email address stored in an ILS. This email address is displayed separately in the profile page (along with e.g. modifiable SMS number).

**Comment by demiankatz** [comment]
Thanks, @EreMaijala, that makes sense too. I think at this point we could conceivably implement either the separate table approach or your design in this PR without creating significant backward compatibility breaks... so that leads to a couple of questions:

1.) Do you think it's worth investing the time into building the "management of multiple email addresses" interface, or is it not useful even if we put in the effort? I'm more than happy to go forward with your proposed solution, and if that meets everyone's needs, that's great... but I'd rather put in the time to "do it right" now if we're going to need the more complex system in future anyway, since changing things at that point will be more expensive than just doing it now.

2.) One significant concern (which is an issue regardless of which approach we take): what happens if the email address changes upstream? It seems like we might want to set a flag somewhere to indicate a user-provided email address, so we can automatically update non-user-administered addresses.

**Comment by EreMaijala** [comment]
1.) I think multiple email addess management might be unnecessarily complicated for most users with little benefit. It's already complex enough to have two email addresses.

2.) Having a flag denoting a user-modified email address would make sense. That wouldn't add too much complexity. Only issue I can think of is the situation where the user requests change of email address in the ILS (using whatever mechanism is available), and its only reflected in VuFind after next login. Then again, that's how it is currently, so I wouldn't consider it a major issue.

**Comment by demiankatz** [comment]
@EreMaijala, I agree, adding a flag should cover the "upstream change" problem, the lack of instant synchronization to VuFind is not a major concern for me, and keeping the user interface simple and streamlined seems a good reason to avoid the more complicated email management system. I'll try to put in those finishing touches in the near future, though I'm not sure if I'll get to it before my vacation next week. I'll flag you in a comment here whenever it gets done, so you can do a final review.

**Comment by mtrojan-ub** [comment]
The verification changes look good.

I also agree with Ere that multiple email addresses per user is most likely too complicated.

btw: Regarding VUFIND-1309, We already have made some customization in IxTheo for SDI, but since the code is already a few years old, very specialized and parts of it are written in C++, it would be very high effort to refactor & contribute that.

also btw / just fyi: We don't use an ILS, but one of our problems is that we run multiple VuFind-instances with different frontends on the same machine+database. Until a few weeks ago (due to the same database), we allowed users to use the same account for both front-ends, resulting in errors because e.g. of favorites in the one frontend where the records haven't been availabile in the other and so on... now we restricted that, but due to the same database the user can't use the same username+email address for a second account in the other frontend.
So either it would be possible to add the frontend/instance as part of the unique key in the user table, or to use separated databases (which leads to higher server maintenance effort).

**Comment by demiankatz** [comment]
Thanks, @mtrojan-ub!

Regarding your problem with multiple instances, I would be inclined to just use separate databases. I think the maintenance burden of managing more than one MySQL database is actually going to be significantly less than the possible problems of sharing a database between instances, especially if there is no particular user benefit to the sharing. If there are database schema changes in future, if you share the database, you will have to simultaneously upgrade all of your VuFind instances to avoid problems -- I think that alone is a strong argument for keeping things separated. In any case, this is a bit off-topic for the PR, so please feel free to email me if you'd like to discuss this further!

**Comment by demiankatz** [comment]
@EreMaijala, I've just added the user_provided_email flag and related behavior as discussed above. Quick tests indicate that it is working, but I definitely need to do more exhaustive testing when time permits.

One more issue that arises: as currently coded, the user is only allowed to change their email address if the authentication method permits. It sounds like you'd rather permit it in all cases, but only pass it along when supported by the authentication method. I think that's an easy change to make -- just wondering if you think there's a need to support any additional configuration (i.e. make it configurable whether to require auth support to allow email, or whether to propagate email changes up to the auth handler, etc.). My gut feeling is that we shouldn't overcomplicate this -- but just checking to see if you have any thoughts on the subject.

**Comment by demiankatz** [comment]
...having reviewed the code, to elaborate on the problem above, the challenge here is that right now, the email change logic is really encapsulated within the Database authentication method. Do we need to move that logic into the base, so that it works across all methods but can be extended/overridden to push things upstream, or do we want to move some of the logic outside of the Auth classes entirely? I think for consistency with other existing code, it probably makes sense to keep it inside Auth, or else we'll have some potentially confusing inconsistent patterns... but there might be arguments to be made for that nonetheless.

**Comment by EreMaijala** [comment]
@demiankatz I'd actually keep the changing of email address in VuFind separate from authentication methods. For database authentication it wouldn't make a difference, but for ILS's it would be a separate function.

**Comment by EreMaijala** [comment]
@mtrojan-ub We work around the duplicate user id problem by prefixing all user id's with an institution code. Related code here:
https://github.com/NatLibFi/NDL-VuFind2/blob/master/module/Finna/src/Finna/Db/Row/User.php#L86
https://github.com/NatLibFi/NDL-VuFind2/blob/master/module/Finna/src/Finna/Db/Table/User.php#L72
https://github.com/NatLibFi/NDL-VuFind2/blob/master/module/Finna/src/Finna/Db/Table/User.php#L114

**Comment by demiankatz** [comment]
@EreMaijala, for now, I've just eliminated email handling in the auth handlers entirely, and moved everything up to the auth manager. I think that will work for now, and if we need to propagate the data upstream in the future, we can figure it out when we have more of a use case.

Now it's time for me to prepare for my vacation -- but I think this is probably in an acceptable form for merging. If you agree, I'll do finishing touches and additional testing when I return from my travels!

**Comment by demiankatz** [comment]
@EreMaijala, just following up to see if you had a chance to review my latest set of changes here.

**Comment by EreMaijala** [comment]
@demiankatz Oops, sorry. I thought I already had. This all looks good to me. One thing that just recently came up when we discussed this and similar functionality: For added security, it might make sense to send a notification about the address being changed to the old email address. This would allow the account owner to react if some kind of account hijacking was attempted.

**Comment by demiankatz** [comment]
@EreMaijala, good idea, I'll work that in before merging.

**Comment by demiankatz** [comment]
@EreMaijala, I had a few minutes at the end of the day to implement the notification you suggested, but I ran out of time before I could test that it actually works. Hopefully I'll get to that tomorrow, but in the meantime, if you want to look at the language strings I added and suggest improvements, I welcome feedback.

**Comment by EreMaijala** [comment]
@demiankatz Works for me. Maybe the notification to the old address could include the (fairly obvious) option to contact the staff if necessary.

**Comment by demiankatz** [comment]
@EreMaijala, yes, the thought did cross my mind about including a message about contacting the staff. My concern was that the method of doing this might vary from institution to institution, and in some cases, contact may not be desired. I wasn't sure how to word it in a way that would be most likely to work for everyone. Maybe the best thing to do is actually to include a note in the comments where the email notification setting is turned on to suggest that users should review the email-specific translation strings and adjust them as needed. By default, perhaps I can include the site-wide email as a contact. I'll experiment and see what I can come up with. Thanks!

**Comment by demiankatz** [comment]
@EreMaijala, I've revised the email message and made some changes to config.ini comments to hopefully clarify the whole situation. I also decided to disable the change_email setting by default, because I think changing email without having verification is dangerous, and turning verification on by default is disruptive, so it seems safer to make the whole stack of options opt-in. Let me know if you have any concerns about this!

I also tested the situation where verification is on, email change is on, and a non-Database authentication method is set. The initial email setting comes in without requiring verification (I think it's better to assume that upstream systems are clean), but subsequent changes do require verification -- I think this is reasonable behavior.

In any case, if you think this is now ready to go, I'll finish up the testing next week and then get it merged.

**Comment by EreMaijala** [comment]
Functionally this looks good to me. The only thing I was wondering if the change email form should already include some instructions. The user will get the message about the verification being sent, but it probably wouldn't hurt to include the information on the form too so it might not get ignored so easily.

**Comment by EreMaijala** [comment]
@demiankatz Yes, something like that.

**Comment by EreMaijala** [comment]
@demiankatz Yes. :)

**Comment by demiankatz** [comment]
I've added a very basic Mink test to cover the new form, but it would be better to account for all of the verification behavior as well. I don't have time at the moment, but I have opened [VUFIND-1351](https://vufind.org/jira/browse/VUFIND-1351) to track this need.

**Pull request merged**